azurelinuxagent/ga/self_update_version_updater.py

```diff
@@ -0,0 +1,181 @@
+# Requires Python 2.6+ and Openssl 1.0+
+
+import datetime
+
+from azurelinuxagent.common import conf, logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.exception import AgentUpgradeExitException, AgentUpdateError
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.version import CURRENT_VERSION
+from azurelinuxagent.ga.ga_version_updater import GAVersionUpdater, VMEnabledRSMUpdates
+
+
+class SelfUpdateType(object):
+    """
+    Enum for different modes of Self updates
+    """
+    Hotfix = "Hotfix"
+    Regular = "Regular"
+
+
+class SelfUpdateVersionUpdater(GAVersionUpdater):
+    def __init__(self, gs_id, last_attempted_manifest_download_time):
+        super(SelfUpdateVersionUpdater, self).__init__(gs_id)
+        self._last_attempted_manifest_download_time = last_attempted_manifest_download_time
+        self._last_attempted_self_update_time = datetime.datetime.min
+
+    @staticmethod
+    def _get_largest_version(agent_manifest):
+        """
+        Get the largest version from the agent manifest
+        """
+        largest_version = FlexibleVersion("0.0.0.0")
+        for pkg in agent_manifest.pkg_list.versions:
+            pkg_version = FlexibleVersion(pkg.version)
+            if pkg_version > largest_version:
+                largest_version = pkg_version
+        return largest_version
+
+    @staticmethod
+    def _get_agent_upgrade_type(version):
+        # We follow semantic versioning for the agent, if <Major>.<Minor>.<Patch> is same, then <Build> has changed.
+        # In this case, we consider it as a Hotfix upgrade. Else we consider it a Regular upgrade.
+        if version.major == CURRENT_VERSION.major and version.minor == CURRENT_VERSION.minor and version.patch == CURRENT_VERSION.patch:
+            return SelfUpdateType.Hotfix
+        return SelfUpdateType.Regular
+
+    def _get_next_upgrade_times(self, now):
+        """
+        Get the next upgrade times
+        return: Next Hotfix Upgrade Time, Next Regular Upgrade Time
+        """
+        def get_next_process_time(last_val, frequency):
+            return now if last_val == datetime.datetime.min else last_val + datetime.timedelta(seconds=frequency)
+
+        next_hotfix_time = get_next_process_time(self._last_attempted_self_update_time,
+                                                 conf.get_self_update_hotfix_frequency())
+        next_regular_time = get_next_process_time(self._last_attempted_self_update_time,
+                                                  conf.get_self_update_regular_frequency())
+
+        return next_hotfix_time, next_regular_time
+
+    def _is_new_agent_allowed_update(self):
+        """
+        This method ensure that update is allowed only once per (hotfix/Regular) upgrade frequency
+        """
+        now = datetime.datetime.now()
+        next_hotfix_time, next_regular_time = self._get_next_upgrade_times(now)
```

Comment:
why do we need to compute both the hotfix and regular times? at any given time, only one of them is valid

Reply:
good point, before we compute I can determine hotfix or regualr version and depending on that compute one time. Makes sense